python/lsst/obs/base/fitsRawFormatterBase.py

```diff
@@ -118,21 +125,82 @@ def makeVisitInfo(self, metadata):
         """
         raise NotImplementedError("Must be implemented by subclasses.")
 
-    def makeWcs(self, metadata):
-        """Construct a SkyWcs from metadata.
+    @abstractmethod
+    def getDetector(self, id):
+        """Return the detector that acquired this raw exposure.
 
         Parameters
         ----------
+        id : `int`
+            The identifying number of the detector to get.
+
+        Returns
+        -------
+        detector : `~lsst.afw.cameraGeom.Detector`
+            The detector associated with that ``id``.
+        """
+        raise NotImplementedError("Must be implemented by subclasses.")
+
+    def makeWcs(self, visitInfo, detector, metadata):
+        """Construct a SkyWcs from the visitInfo and detector, or metadata
+        of an exposure.
+
+        If the boresight or camera geometry do not support this method of
+        WCS creation, this falls back on the header metadata-based version
+        (typically a purely linear FITS crval/crpix/cdmatrix WCS).
+
+        Parameters
+        ----------
+        visitInfo : `~lsst.afw.image.VisitInfo`
+            Description
+        detector : `~lsst.afw.cameraGeom.Detector`
+            Description
         metadata : `~lsst.daf.base.PropertyList`
             Header metadata.  May be modified in-place.
 
         Returns
         -------
-        wcs : `~lsst.afw.geom.SkyWcs`
+        skyWcs : `~lsst.afw.geom.SkyWcs`
             Reversible mapping from pixel coordinates to sky coordinates.
         """
-        from lsst.afw.geom import makeSkyWcs
-        return makeSkyWcs(metadata, strip=True)
+        skyWcs = self._createSkyWcsFromMetadata(metadata)
+
+        log = lsst.log.Log.getLogger("fitsRawFormatter")
+        if visitInfo is None:
+            msg = "No VisitInfo; cannot access boresight information. Defaulting to metadata-based SkyWcs."
+            log.warn(msg)
+            return skyWcs
+        try:
+            skyWcs = createInitialSkyWcs(visitInfo, detector)
+        except InitialSkyWcsError as e:
+            msg = "Cannot create SkyWcs using VisitInfo and Detector, using metadata-based SkyWcs: %s"
```

Comment:
I'm a bit torn (for both Gen2 and Gen3) whether this should be an automatic fallback with a warning vs. a hard failure.  If we start configuring our pipelines (and ingest, which also uses initial WCSs in Gen3 to generate regions-on-sky) to expect good initial WCSs of the sort that requires a boresight+cameraGeom WCS (at least for some obs packages), we're just delaying an astrometry error and making it harder to debug.

Could we find a way to let obs package overrides of the Formatter (or Mapper, for Gen2) turn this into a hard failure?